### ISSUE TITLE
Issues with gffutils.create_db attribute ordering

### ISSUE THREAD

**Issue by cbrueffer** [daler/gffutils#79]
Hi,

I'm debugging a problem in the cloudbiolinux (https://github.com/chapmanb/cloudbiolinux/) script scripts/prepare_tx_gff.py. I have created a minimal test case and input GTF in the attached ZIP file. Since I'm not sure whether this should be fixed inside gffutils or in the consumer, I try my luck here first :-)

I have tested both gffutils 0.8.7.1 as well as the current version from git.

The problem is that prepare_tx_gff.py writes a GTF file with attributes that have gene_id and transcript_id shuffled into the middle of the attributes string, instead of beginning with them. The input GTF file (minimal.gtf in the attached archive is the first line of my actual GTF file) however starts the attribute string correctly with gene_id and transcript_id. This, in turn, leads to gtfToGenePred from the UCSC tools erroring out later on.

The gffutils DB is created like so:

```
    gffutils.create_db(gtf, dbfn=db_file,
                       disable_infer_genes=False,
                       disable_infer_transcripts=False,
                       id_spec={"gene": "gene_id", "transcript": "transcript_id"},
                       merge_strategy="create_unique",
                       keep_order=True,
                       verbose=True)
```

If I read the docs correctly, ```keep_order=True``` should lead to the attributed being stored in exactly the same sequence as they appear in the input file, however this is not the case if you then iterate through the DB:

```
for feature in db.all_features():
            out_handle.write(str(feature) + "\n")
```

The output lines look like this:

```
chr1	hg38_knownGene	exon	11874	12227	0	+	.	knownCanonicalChrom "NA"; description "Homo sapiens DEAD/H (Asp-Glu-Ala-Asp/His) box helicase 11 like 1 (DDX11L1), non-coding RNA."; protAcc; knownCanonicalChromStart "NA"; knownToRefSeq "NR_046018"; knownIsoformsClusterID "1"; gene_id "uc001aaa.3"; tss_id "TSS1"; alias "DDX11L1"; transcript_id "uc001aaa.3"; mRNA "NR_046018"; knownCanonicalChromEnd "NA"; knownCanonicalClusterId "NA"; refseq "NR_046018"; locusLink "100287102"; geneSymbol "DDX11L1";
chr1	gffutils_derived	transcript	11874	12227	.	+	.	transcript_id "uc001aaa.3"; gene_id "uc001aaa.3";
```

I should add that the code in the ZIP archive is boiled down from the actual prepare_tx_gff.py in cloudbiolinux.

Any insight would be appreciated!

[bugtest.zip](https://github.com/daler/gffutils/files/417193/bugtest.zip)

**Comment by cbrueffer** [comment]
CCing @chapmanb and @roryk on this as cloudbiolinux people.

**Comment by roryk** [comment]
Hi Christian,

Does the original GTF file not have the `gene` and `transcript` entries? The call to `create_db` looks like it doesn't because the infer extent is not turned off. I think what is happening is gffutils is inferring the `gene` and `transcript` entries, but since they aren't in the original GTF file, they don't get put in the right order. I'm not sure how to get the ordering to work properly here, we might need to iterate over all gene features and then all transcript features and then the rest of the features instead of just going over all features.

**Comment by cbrueffer** [comment]
Hi Rory,

thanks for the comment! You mean ```gene``` and ```transcript``` in the third column of the GTF? No, the only entries appearing are CDS, exon, start_codon and stop_codon. And indeed, the call to ```guess_disable_infer_extent``` in ```prepare_tx_gff.py``` results in False for both gene and transcript (hence the two False in the example above).

**Comment by roryk** [comment]
Hi Christian,

Yeah exactly-- those are getting inferred and then not getting spit out in the right order. I'm not 100% sure how to handle this appropriately. Things will end up being out of coordinate order no matter what we do. What are the errors that come up for the downstream tools? We might need to strip those gene/transcript features before running them.

**Comment by cbrueffer** [comment]
Hmm, not sure I understand the problem correctly. If you look at the minimal.gtf file, the entry looks like this:

```
chr1	hg38_knownGene	exon	11874	12227	0	+	.	gene_id "uc001aaa.3"; transcript_id "uc001aaa.3"; geneSymbol "DDX11L1"; mRNA "NR_046018"; refseq "NR_046018"; protAcc ""; description "Homo sapiens DEAD/H (Asp-Glu-Ala-Asp/His) box helicase 11 like 1 (DDX11L1), non-coding RNA."; locusLink "100287102"; alias "DDX11L1"; knownToRefSeq "NR_046018"; knownIsoformsClusterID "1"; knownCanonicalChrom "NA"; knownCanonicalChromStart "NA"; knownCanonicalChromEnd "NA"; knownCanonicalClusterId "NA"; tss_id "TSS1";
```

So, with ```keep_order=True``` the entry ordering should stay stable, but it doesn't when looking at the output above. The inferred transcript line is correct.

The attribute ordering issue is the reason ```gtfToGenePred``` chokes, details here: https://groups.google.com/a/soe.ucsc.edu/forum/#!topic/genome/Y4IYcaYy7rc

**Comment by cbrueffer** [comment]
Reading the comments again, I think we refer to different ordering issues. My problem is the attributes order (last GTF column), whereas I think you refer to the order of the two output lines in the original post?

**Comment by roryk** [comment]
Ohhhhhh, I see. We might have to spit out a cleaned up GTF file then to get that to work.

**Comment by daler** [comment]
Thanks for the test script and example file.  I wasn't able to reproduce the `keep_order` issue, but I did notice that the value for `protAcc` was an empty string but in the output the value was removed (so it was only the key with no value). This was the expected behavior (I think?) for GFF files (see [this example file](http://daler.github.io/gffutils/examples.html#glimmer-nokeyval-gff3) but GTF empty fields should probably have quotes.

8aafedfd5f2a includes a test that verifies the output exon feature matches the input exon feature, and also verifies the empty GTF value shows up as an empty double-quoted string.

Does the latest git version help the downstream issue?
